File: ios/Plugin/CPUMonitoringPlugin.m
#import <Foundation/Foundation.h>
#import <Capacitor/Capacitor.h>

// Define the plugin using the CAP_PLUGIN Macro, and
// each method the plugin supports using the CAP_PLUGIN_METHOD macro.
CAP_PLUGIN(CPUMonitoringPlugin, "CPUMonitoring",
           CAP_PLUGIN_METHOD(getLoad, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(setLoadRefreshTime, CAPPluginReturnNone);
)

File: ios/Plugin/CPUMonitoringPlugin.swift
import Foundation
import Capacitor

@objc(CPUMonitoringPlugin)
public class CPUMonitoringPlugin: CAPPlugin {
    private let implementation = CPUMonitoring()
    
    @objc override public func load() {
        implementation.startLoadLoop()
        self.initEvents()
    }

    @objc func getLoad(_ call: CAPPluginCall) {
        do {
            call.resolve(["load": try implementation.getLoad()])
        } catch {
            call.reject((error as! CustomError).message,
                            (error as! CustomError).code.rawValue,
                            (error as! CustomError).error)
        }
    }
    
    @objc func setLoadRefreshTime(_ call: CAPPluginCall) {
        guard let interval = call.getInt("interval") else {
            return call.reject("setLoadRefreshTime:: intervalMillis not provided")
        }
        
        implementation.setLoadRefreshTime(interval)
        call.resolve()
    }
    
    private func initEvents() {
        NotificationCenter.default.addObserver(self, selector: #selector(self.notifyCustomListeners(notification:)), name: Notification.Name("CPULoad"), object: nil)
    }
    
    @objc private func notifyCustomListeners(notification: Notification) {
        self.notifyListeners(notification.name.rawValue, data: notification.userInfo as? [String: Any])
    }
}
